File: 908-MiddleOfTheLinkedList/908-MiddleOfTheLinkedList.c
// Last updated: 9/14/2025, 9:12:00 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* middleNode(struct ListNode* head){
      int count=0;
      struct ListNode *temp=head;
      while(temp!=NULL){
          count++;
          temp=temp->next;
      }
      temp=head;
      int mid=count/2;
      for(int i=0;i<mid;i++){
           temp=temp->next;
      }
      return temp;
}

File: 908-MiddleOfTheLinkedList/908-MiddleOfTheLinkedList.java
// Last updated: 9/14/2025, 9:11:57 AM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // tortoise and hare approach
        ListNode tortoise = head;
        ListNode hare = head;
        while(hare!=null && hare.next!=null){
            tortoise = tortoise.next;
            hare = hare.next.next;
        }
        //since if hare covers d distance tortoise cover d/2 distance
        return tortoise;
    }
}